### PR TITLE
feat(lsp): lsp js e2e

### PR DESCRIPTION
**Pull request by ajbt200128**
This PR wires up some LSP/VSCODE packaging stuff together with LSP.js so it can actually run in the editor.

## Test plan

there are tests written in js/language_server/src/tests

### CODE REVIEW

**Comment by ajbt200128** [comment]
TODO: figure out artifact stuff

File: .github/workflows/build-test-javascript.jsonnet [hunk 1]
```diff
@@ -134,17 +134,32 @@ local test_job = {
     {
       name: 'Test JS artifacts',
       run: |||
-        make -C js -j $(nproc) test
-        make -C js/tests
```
Comment (tpetr) [inline .github/workflows/build-test-javascript.jsonnet:140]: was this removal intentional?

File: js/language_server/src/tests/runTests.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import * as path from "path";
+import find = require("find-process");
+
+import { runTests } from "@vscode/test-electron";
+
+async function go() {
```
Comment (tpetr) [inline js/language_server/src/tests/runTests.ts:6]: there isn't a cleaner way to get xvfb's DISPLAY?

File: js/language_server/src/server.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+connection.onInitialize(async (params) => {
+  // this is kind of gross to do this here, but it's simple.
+  server = await LSFactory();
+  // Tell the LS to use this function to write to stdout
+  server.setWriteRef((json) => {
+    // Parse JSON from ocaml code
+    const packet = JSON.parse(json);
+    // extract method and params since that's what the LSP expects
+    const method = packet.method;
+    const params = packet.params;
```
Comment (tpetr) [inline js/language_server/src/server.ts]: if you wanna be a cool javascripter you could do:
```
const { method, params } = packet;
```

File: js/engine/Makefile.include [hunk 4]
```diff
@@ -1,7 +1,8 @@
 EMCC_DEFAULTS = \
    -sALLOW_MEMORY_GROWTH=1 \
    -sEXPORTED_RUNTIME_METHODS=UTF8ToString,stringToUTF8,stringToAscii,lengthBytesUTF8,getValue,setValue,intArrayFromString,writeArrayToMemory \
-   -sMODULARIZE
+   -sMODULARIZE \
+   -sSINGLE_FILE=1
```
Comment (tpetr) [inline js/engine/Makefile.include:5]: I have flip-flopped on this a few times in the past -- what's your reasoning for embedding?

**Review by tpetr — APPROVED** [review]
LGTM, just some nits and questions. The packaging situation is pretty gross but let's get this merged and we can do packaging in a separate PR

File: .github/workflows/build-test-javascript.jsonnet (hunk 1 above)
Comment (ajbt200128) [reply .github/workflows/build-test-javascript.jsonnet:140]: Yea the makefile had js/tests in it already so this was doing double duty

File: js/language_server/src/tests/runTests.ts (hunk 2 above)
Comment (ajbt200128) [reply js/language_server/src/tests/runTests.ts:6]: I don't think so, this was copy pasted from the official vscode language server repo 🤷 so

File: js/engine/Makefile.include (hunk 4 above)
Comment (ajbt200128) [reply js/engine/Makefile.include:5]: When in node the wasm loading doesn't work, I'm not 100% sure why but I think it's something to do with how the vscode extensions get packed. Also this is a lot easier to work with imo

File: js/language_server/src/server.ts (hunk 3 above)
Comment (ajbt200128) [reply js/language_server/src/server.ts]: noted for next time, but too lazy to fix rn

**Comment by tpetr** [comment]
Any codeowners wanna ship this so that we can keep rolling?

File: .github/workflows/build-test-javascript.jsonnet [hunk 5]
```diff
@@ -134,17 +134,32 @@ local test_job = {
...
+          make -C js -j $(nproc) test
       |||
     },
+    {
+      name: 'Setup APT for xvfb',
```
Comment (aryx) [inline .github/workflows/build-test-javascript.jsonnet:149]: maybe a small comment here explaining what is xvfb, and why the hell we need libgtk-3.0 in semgrep!

Also maybe those things would be better in the Makefile, like Make install-deps-UBUNTU-for-lsp-js

File: js/engine/Makefile.include (hunk 4 above)
Comment (aryx) [reply js/engine/Makefile.include:5]: maybe good to add a comment here explaining the history and why we need those flags, so the next person can understand why we have this very specific config

File: js/language_server/dune [hunk 6]
```diff
@@ -1,6 +1,8 @@
 (env
  (js_of_ocaml (flags -no-inline -debug-info)))
 
+(data_only_dirs node_modules)
```
Comment (aryx) [inline js/language_server/dune:4]: you can also do instead the complement with (dirs src tests)

File: js/language_server/fixtures/test.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+if 5 == 5:
```
Comment (aryx) [inline js/language_server/fixtures/test.py]: fixtures?

File: js/language_server/package.json [hunk 8]
```diff
@@ -1,37 +1,68 @@
 {
-  "name": "@semgrep/engine",
+  "name": "@semgrep/language-server",
   "version": "0.0.1",
-  "description": "Javascript bindings for the semgrep engine",
-  "main": "dist/index.js",
+  "description": "Javascript bindings for the LSP",
+  "main": "extension.js",
```
Comment (aryx) [inline js/language_server/package.json]: I really don't like this "extension" terminology ... it's very vague, when you say "semgrep extensions", why
someone would think about vscode plugins? we should call it semgrep-lsp,  or editor extensions, or editor plugin.

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
test plan: ?

**Comment by aryx** [comment]
Can I install the stuff produced in Vscode? How do I run it? How do I test/exercise the code and artifacts added in this PR?

File: .github/workflows/build-test-javascript.jsonnet (hunk 5 above)
Comment (ajbt200128) [reply .github/workflows/build-test-javascript.jsonnet:149]: All these deps are for xvfb, which is essentially a virtual (x11) display, so VSCode tests can run, since they require a display. Unlike emacs vscode is lame and does not have a daemon mode

File: js/language_server/dune (hunk 6 above)
Comment (ajbt200128) [reply js/language_server/dune:4]: will switch to that

File: js/language_server/fixtures/test.py (hunk 7 above)
Comment (ajbt200128) [reply js/language_server/fixtures/test.py]: should probably rename to targets not fixtures, to stay consistent

File: js/language_server/package.json (hunk 8 above)
Comment (ajbt200128) [reply js/language_server/package.json]: Good call out, I will rename this

**Comment by ajbt200128** [comment]
test plan is that we use the vscode testing architecture to ensure it works. There'll be another PR for proper packaging for our extension, but to use it in vscode you'd need to take a look at js/language_server/src/tests/integration.test.ts.

File: .github/workflows/build-test-javascript.jsonnet (hunk 5 above)
Comment (ajbt200128) [reply .github/workflows/build-test-javascript.jsonnet:149]: but the deps are specifcally for testing in a headless ubuntu CI, not anywhere else, so I think I'd prefer to leave them in this workflow since that's the only time they matter, no need to abstract

Comment (ajbt200128) [reply .github/workflows/build-test-javascript.jsonnet:149]: or sorry, the other deps are for some testing framework that are used under the hood, let me think about it

File: js/language_server/dune (hunk 6 above)
Comment (ajbt200128) [reply js/language_server/dune:4]: actually prefer the data_only_dirs since Main.ml is top level
